### PR TITLE
newnode 2.1.4 (new formula)

### PR DESCRIPTION
**Pull request by flyinhome**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

### CODE REVIEW

**Review by flyinhome — COMMENTED** [review]
adds dependency on macos high sierra so it won't try to build on linux

**Comment by SMillerDev** [comment]
> adds dependency on macos high sierra so it won't try to build on linux

Why shouldn't it?

File: Formula/n/newnode-helper.rb
Comment (SMillerDev) [inline Formula/n/newnode-helper.rb:51]: If this isn't provided by upstream we shouldn't ship it

**Review by SMillerDev — CHANGES_REQUESTED** [review]
Way too much magic going on in this install

**Comment by flyinhome** [comment]
On 11/12/23 06:17, Sean Molenaar wrote:

> If this isn't provided by upstream we shouldn't ship it

Since the git repo is ours, we can fix the git repo to provide this plist.

Keith

**Comment by flyinhome** [comment]
> > adds dependency on macos high sierra so it won't try to build on linux
> 
> Why shouldn't it?

Separately from the effort to make newnode-helper easily installable via homebrew, we've set up a snap to facilitate easy installation on linux platforms.  We judged (perhaps incorrectly) snap to be more likely to facilitate installs on a wide variety of Linux platforms, and homebrew to be more likely to facilitate installs for MacOS because homebrew is very widely used on MacOS.

We could adapt this homebrew formula to facilitate homebrew installs on Linux also, since the C source code for newnode builds fine on Linux.  But at least in my experience things can get confusing for users when the same software package can be installed via multiple mechanisms.  For example, when version X of the software is installed via one mechanism and upgraded to version Y via a different mechanism.  

So if we wanted to facilitate installs of newnode-helper via both homebrew and snap, we'd want to make sure that all future upgrades (by whatever mechanism) left the resulting installation in a consistent state no matter which mechanism was  used to do the upgrade.  And that seems tricky, and at least, not what we wanted to try on first attempt.

**Comment by flyinhome** [comment]
Just to let people know, the authors' intention for the newnode-helper formula is to update the underlying repo to put more of the "magic" in the newnode repository code, and less of the "magic" in the homebrew formula, and submit another PR once the new combination is tested.  

In some ways this is more complicated because it requires coordination among more parties and makes testing more difficult.  But we understand that homebrew maintainers want to keep formulae as simple and straightforward as possible.

**Comment by flyinhome** [comment]
> > We could adapt this homebrew formula to facilitate homebrew installs on Linux also, since the C source code for newnode builds fine on Linux. But at least in my experience things can get confusing for users when the same software package can be installed via multiple mechanisms.
> 
> That seems like the best solution since the current split seems very arbitrary. I see the risk of https://xkcd.com/1987/, but that's up to users to fix. Not up to us to restrict.

The split isn't arbitrary, since MacOS and Linux are very different platforms, and most MacOS users are accustomed to getting their software from different sources than most Linux users. 

Also, what we need for newnode-helper is for the app to run as a daemon full time, and as its own user (least privilege).   There's no portable way to do that on Linux.   Snap provides a way to do that across a wide variety of Linux platforms.    If homebrew on Linux supports the ability to do that on a large number of Linux platforms, please let me know what it is.  But I'm under the impression that homebrew really isn't designed for that use case, and it takes "magic" to get it to work even on a Mac.

**Comment by SMillerDev** [comment]
> The split isn't arbitrary, since MacOS and Linux are very different platforms, and most MacOS users are accustomed to getting their software from different sources than most Linux users.

It's arbitrary because the tool works for both, but as a package manager we're disabling that choice. It's like buying a car in two countries and in one of them the car will refuse to drive on highways.

> Also, what we need for newnode-helper is for the app to run as a daemon full time, and as its own user (least privilege). There's no portable way to do that on Linux. Snap provides a way to do that across a wide variety of Linux platforms. If homebrew on Linux supports the ability to do that on a large number of Linux platforms, please let me know what it is. But I'm under the impression that homebrew really isn't designed for that use case, and it takes "magic" to get it to work even on a Mac.

https://docs.brew.sh/Formula-Cookbook#service-files

**Comment by fxcoudert** [comment]
> Just to let people know, the authors' intention for the newnode-helper formula is to update the underlying repo to put more of the "magic" in the newnode repository code, and less of the "magic" in the homebrew formula, and submit another PR once the new combination is tested.

Yes, that's definitely the way to go forward. I am closing this inactive PR, but we'll be happy to consider a new one once the install is streamlined. Thanks @flyinhome !

**Pull request closed without merging**
